File: aoc2024/aoc3.hs
import System.Environment (getArgs)
import Data.Char (isDigit)
import Text.ParserCombinators.ReadP
import Data.Functor (($>))
import Control.Monad.State

data Instruction = Do | Dont | Mul Int Int deriving Show

newtype InstructionList = IL [Instruction] deriving Show

parseInstr :: ReadP Instruction
parseInstr = (string "do()" $> Do) +++ (string "don't()" $> Dont)
             +++ (Mul <$> (string "mul(" *> (read <$> many1 (satisfy isDigit)))
                      <*> (string "," *> (read <$> many1 (satisfy isDigit))) <* string ")")

parseIL :: ReadP InstructionList
parseIL = IL <$> parsy
  where
    parsy = concat <$> manyTill (((:[]) <$> parseInstr) <++ (satisfy (const True) $> [])) eof

eval1 :: InstructionList -> Int
eval1 (IL instrs) = sum $ doer <$> instrs
  where
    doer Do = 0
    doer Dont = 0
    doer (Mul a b) = a * b

eval2 :: InstructionList -> Int
eval2 (IL instrs) = snd $ execState (mapM_ doer instrs) (True, 0)
  where
    doer :: Instruction -> State (Bool, Int) ()
    doer Do = modify (\s -> (True, snd s))
    doer Dont = modify (\s -> (False, snd s))
    doer (Mul a b) = modify (\s -> (fst s, snd s + if fst s then a*b else 0))

main :: IO ()
main = do
    args <- getArgs
    let filename = if null args then "aoc3.in" else head args
    mydata <- readFile filename
    let myParser = readP_to_S parseIL
    let myIL = fst $ head $ filter (null . snd) $ myParser mydata
    putStr "Part 1: "
    print $ eval1 myIL
    putStr "Part 2: "
    print $ eval2 myIL

File: aoc2024/aoc3.py
from aoc_util import get_data
import re

data = get_data(3)
total = 0
for m in re.finditer(r"mul\((\d+), *(\d+)\)", data):
    total += int(m.group(1)) * int(m.group(2))
print("Part 1:", total)


total = 0
enabled = True
for m in re.finditer(r"(do\(\))|(don't\(\))|mul\((\d+), *(\d+)\)", data):
    if m.group(1):
        enabled = True
    elif m.group(2):
        enabled = False
    else:
        if enabled:
            total += int(m.group(3)) * int(m.group(4))
print("Part 2:", total)
